app/models/work_package/validations.rb

```diff
@@ -129,13 +152,22 @@ def validate_category
     end
   end
 
-  def validate_children
-    children.select { |c| !c.valid? }.each do |child|
-      child.errors.each do |attribute, message|
-        full_message = child.errors.full_message(attribute, message)
-        errors.add(:base, "#{I18n.t('label_child_element')} ##{child.id}: #{full_message}")
-      end
+  def validate_descendants
+    all_descendants = descendants.eager_load_for_validation
+
+    # As the current state of the work package may differ from the
+    # db. We assign the already instantiated work package with all
+    # it's changes as the parent object to the work package's direct
+    # children. Other descendants should not have altered parents.
+    all_descendants.each do |wp|
+      wp.parent = self if wp.parent_id == id
     end
+
+    invalid_descendants = all_descendants.select { |c|
```

Comment:
This is probably negligible compared to the SQL performance improvements, but why do you select invalid children here, only to loop over them later again? It may not be prettier, but faster and without intermediary arrays to simple loop over all_descendants, collecting errors as you go?